scripts/3rd-party/backend.js

```diff
@@ -0,0 +1,386 @@
+ * @property {?SizeVariant} thumb2x
+ * @property {?SizeVariant} thumb
+ */
+
+/**
+ * @typedef SizeVariant
+ *
+ * @property {number} type
+ * @property {string} url
+ * @property {number} width
+ * @property {number} height
+ * @property {number} filesize
+ */
+
+/**
+ * @typedef SortingCriterion
+ *
+ * @property {string} column
+ * @property {string} order
+ */
+
+/**
+ * @typedef Album
+ *
+ * @property {string}  id
+ * @property {string}  parent_id
+ * @property {string}  created_at
+ * @property {string}  updated_at
+ * @property {string}  title
+ * @property {?string} description
+ * @property {string}  license
+ * @property {Photo[]} photos
+ * @property {Album[]} [albums]
+ * @property {?string} cover_id
+ * @property {?Thumb}  thumb
+ * @property {string}  [owner_name] optional, only shown in authenticated mode
+ * @property {boolean} is_public
+ * @property {boolean} is_downloadable
+ * @property {boolean} is_share_button_visible
+ * @property {boolean} is_nsfw
+ * @property {boolean} grants_full_photo
+ * @property {boolean} requires_link
+ * @property {boolean} has_password
+ * @property {boolean} has_albums
+ * @property {?string} min_taken_at
+ * @property {?string} max_taken_at
+ * @property {?SortingCriterion} sorting
+ */
+
+/**
+ * @typedef TagAlbum
+ *
+ * @property {string}   id
+ * @property {string}   created_at
+ * @property {string}   updated_at
+ * @property {string}   title
+ * @property {?string}  description
+ * @property {string[]} show_tags
+ * @property {Photo[]}  photos
+ * @property {?Thumb}   thumb
+ * @property {string}   [owner_name] optional, only shown in authenticated mode
+ * @property {boolean}  is_public
+ * @property {boolean}  is_downloadable
+ * @property {boolean}  is_share_button_visible
+ * @property {boolean}  is_nsfw
+ * @property {boolean}  grants_full_photo
+ * @property {boolean}  requires_link
+ * @property {boolean}  has_password
+ * @property {?string}  min_taken_at
+ * @property {?string}  max_taken_at
+ * @property {?SortingCriterion}  sorting
+ * @property {boolean}  is_tag_album always true
+ */
+
+/**
+ * @typedef SmartAlbum
+ *
+ * @property {string}  id
+ * @property {string}  title
+ * @property {Photo[]} photos
+ * @property {?Thumb}  thumb
+ * @property {boolean} is_public
+ * @property {boolean} is_downloadable
+ * @property {boolean} is_share_button_visible
+ */
+
+/**
+ * @typedef Thumb
+ *
+ * @property {string}  id
+ * @property {string}  type
+ * @property {string}  thumb
+ * @property {?string} thumb2x
+ */
+
+/**
+ * @typedef SharingInfo
+ *
+ * DTO returned by `Sharing::list`
+ *
+ * @property {{id: number, album_id: string, user_id: number, username: string, title: string}[]} shared
+ * @property {{id: string, title: string}[]}                                                      albums
+ * @property {{id: number, username: string}[]}                                                   users
+ */
+
+/**
+ * @typedef SearchResult
+ *
+ * DTP returned by `Search::run`
```

Comment:
From the perspective of the controller in the backend you are right. However, at the end of the Laravel pipeline we sent JSON objects to the client.

There is no syntactically enforced equality between what the backend returns and what is documented in this file. They must be kept in sync manually. I only wrote this file for a) documentation purposes of the API and b) because it gives me auto-completion in my IDE. It also allows SonarCloud to analyze our code much more thoroughly what immediately led to an D rating.